table/block_based/block_based_table_iterator.cc

```diff
@@ -672,65 +701,144 @@ void BlockBasedTableIterator::FindReadAheadSizeUpperBound() {
+  // scenarios
+  if (read_curr_block) {
+    if (!DoesContainBlockHandles()) {
+      // Scenario 1 : read_curr_block (callback made on miss block which caller
+      //              was reading) and it has no existing handles in queue. i.e.
+      //              index_iter_ is pointing to block that is being read by
+      //              caller.
+      //
+      // Add current block here as it doesn't need any lookup.
+      BlockHandleInfo block_handle_info;
+      block_handle_info.handle_ = index_iter_->value().handle;
+      block_handle_info.SetFirstInternalKey(
+          index_iter_->value().first_internal_key);
+
+      end_updated_offset = block_handle_info.handle_.offset() + footer +
+                           block_handle_info.handle_.size();
+      block_handles_.emplace_back(std::move(block_handle_info));
+
+      index_iter_->Next();
+      is_index_at_curr_block_ = false;
+      found_first_miss_block = true;
+    } else {
+      // Scenario 2 : read_curr_block (callback made on miss block which caller
+      //              was reading) but the queue already has some handles.
+      //
+      // It can be due to reading error in second buffer in FilePrefetchBuffer.
+      // BlockHandles already added to the queue but there was error in fetching
+      // those data blocks. So in this call they need to be read again.
+      assert(block_handles_.front().is_cache_hit_ == false);
+      found_first_miss_block = true;
+      // Initialize prev_handles_size to 0 as all those handles need to be read
+      // again.
+      prev_handles_size = 0;
+      start_updated_offset = block_handles_.front().handle_.offset();
+      end_updated_offset = block_handles_.back().handle_.offset() + footer +
+                           block_handles_.back().handle_.size();
+    }
+  } else {
+    // Scenario 3 : read_curr_block is false (callback made to do additional
+    //              prefetching in buffers) and the queue already has some
+    //              handles from first buffer.
+    if (DoesContainBlockHandles()) {
+      start_updated_offset = block_handles_.back().handle_.offset() + footer +
+                             block_handles_.back().handle_.size();
+      end_updated_offset = start_updated_offset;
+    } else {
+      // Scenario 4 : read_curr_block is false (callback made to do additional
+      //              prefetching in buffers) but the queue has no handle
+      //              from first buffer.
+      //
+      // It can be when Reseek is from block cache (which doesn't clear the
+      // buffers in FilePrefetchBuffer but clears block handles from queue) and
+      // reseek also lies within the buffer. So Next will get data from
+      // exisiting buffers untill this callback is made to prefetch additional
+      // data. All handles need to be added to the queue starting from
+      // index_iter_.
+      assert(index_iter_->Valid());
+      start_updated_offset = index_iter_->value().handle.offset();
```

Comment:
Would this cause some read amplification? The previous readahead might have read the block partially (end_offset > index_iter_->value().handle.offset()), but we will read it again here.

Reply:
No it shouldn't. We are not reading the data blocks again. It only iterates over the index_iter_ and add block_handles from index_iter_ uptill end offset.